File: cpp/2021/mar/word-subsets.cc
#include <vector>
#include <string>

using namespace std;

class Solution
{
public:
    vector<string> wordSubsets(vector<string> &A, vector<string> &B)
    {
        vector<int> count(26), tmp(26);
        int i;
        for (string b : B)
        {
            tmp = counter(b);
            for (i = 0; i < 26; ++i)
            {
                count[i] = max(count[i], tmp[i]);
            }
        }
        vector<string> ret;
        for (auto cb = A.cbegin(); cb != A.cend(); ++cb)
        {
            string s = *cb;
            tmp = counter(s);
            for (i = 0; i < 26; ++i)
            {
                if (tmp[i] < count[i])
                {
                    break;
                }
            }
            if (i == 26)
            {
                ret.push_back(s);
            }
        }
        return ret;
    }

private:
    vector<int> counter(string &word)
    {
        vector<int> count(26);
        for (char c : word)
        {
            count[c - 'a']++;
        }
        return count;
    }
};

File: code/python/2021/mar/word-subsets.py
from typing import List, Counter
import operator
from functools import reduce
import collections


class Solution:
    def wordSubsets(self, A: List[str], B: List[str]) -> List[str]:
        c_b = reduce(operator.or_, (Counter(w) for w in B))
        return [a for a in A if c_b & Counter(a) == c_b]

    def wordSubsets_(self, A, B):
        count = collections.Counter()
        for b in B:
            count |= collections.Counter(b)
        return [a for a in A if not count - Counter(a)]